### PR TITLE
fs, win: add support for user symlinks - take 2

### PR DESCRIPTION
**Pull request by bzoz**
With https://github.com/libuv/libuv/pull/1346 being open so long, this is another take on the subject. This time we will use `SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE` by default. If `CreateSymbolicLink` fails with `ERROR_INVALID_PARAMETER`, we will disable the flag and retry.

Fixes: https://github.com/libuv/libuv/issues/1157

### CODE REVIEW

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/369/

File: src/win/fs.c [hunk 1]
```diff
@@ -1701,21 +1702,43 @@ static void fs__symlink(uv_fs_t* req) {
...
-    }
-  } else {
+    return;
+  }
+  if (!pCreateSymbolicLinkW) {
     SET_REQ_UV_ERROR(req, UV_ENOSYS, ERROR_NOT_SUPPORTED);
+    return;
+  }
+
+  if (flags & UV_FS_SYMLINK_DIR) {
```
Comment (cjihrig) [inline src/win/fs.c]: Can you drop the braces on this `if-else`

File: src/win/fs.c [hunk 2]
```diff
@@ -1701,21 +1702,43 @@ static void fs__symlink(uv_fs_t* req) {
...
+    flags = SYMBOLIC_LINK_FLAG_DIRECTORY;
+  } else {
+    flags = uv__file_symlink_usermode_flag;
+  }
+
+  result = pCreateSymbolicLinkW(new_pathw, pathw, flags);
+  if (result != 0) {
```
Comment (cjihrig) [inline src/win/fs.c]: Very minor, but if you inline the `pCreateSymbolicLinkW()` call here, you can get rid of `result` completely.

File: src/win/fs.c [hunk 3]
```diff
@@ -1701,21 +1702,43 @@ static void fs__symlink(uv_fs_t* req) {
...
+   * symlinks.
+   */
+  err = GetLastError();
+  if (err == ERROR_INVALID_PARAMETER &&
+      flags & SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE) {
+    /* This system does not support user-mode symlinks. We will clear the
+     * unsupported flag and retry.
+     */
+    uv__file_symlink_usermode_flag = 0;
+    fs__symlink(req);
```
Comment (cjihrig) [inline src/win/fs.c]: Is it hypothetically possible for `fs__symlink()` to keep calling itself? Maybe add a check that `uv__file_symlink_usermode_flag` isn't already zero?

**Comment by bzoz** [comment]
@cjihrig updated, PTAL.

`fs__symlink` will not call itself. We set `uv__file_symlink_usermode_flag` to zero before retrying. I've slightly refactored the code, it should be more clearly visible now.

File: src/win/fs.c [hunk 4]
```diff
@@ -1699,23 +1700,42 @@ static void fs__create_junction(uv_fs_t* req, const WCHAR* path,
 static void fs__symlink(uv_fs_t* req) {
   WCHAR* pathw = req->file.pathw;
   WCHAR* new_pathw = req->fs.info.new_pathw;
-  int flags = req->fs.info.file_flags;
-  int result;
+  int flags;
```
Comment (santigimeno) [inline src/win/fs.c]: nit: while you're here you could separate declaration from assignment.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/386/

File: src/win/fs.c (hunk 4 above)
Comment (bzoz) [inline src/win/fs.c]: Sure

**Comment by bzoz** [comment]
Unrelated failure on RedHat:
```
not ok 181 - hrtime
# exit code 134
# Output from process `hrtime`:
# Assertion failed in test/test-hrtime.c on line 50: diff < (uint64_t) 80 * NANOSEC / MILLISEC
```

**Comment by bzoz** [comment]
Rebased, last CI: https://ci.nodejs.org/job/libuv-test-commit/387/

**Comment by bzoz** [comment]
landed in https://github.com/libuv/libuv/commit/89d31932a595fdec18dd19b70142816312ea2338

File: src/win/fs.c [hunk 5]
```diff
@@ -1713,25 +1714,46 @@ static void fs__create_junction(uv_fs_t* req, const WCHAR* path,
...
+  /* Something went wrong. We will test if it is because of user-mode
+   * symlinks.
+   */
+  err = GetLastError();
+  if (err == ERROR_INVALID_PARAMETER &&
+      flags & SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE) {
+    /* This system does not support user-mode symlinks. We will clear the
+     * unsupported flag and retry.
+     */
+    uv__file_symlink_usermode_flag = 0;
```
Comment (refack) [inline src/win/fs.c]: Question: is this a common pattern, to use a global var for caching state?
Obviously it's simpler than https://github.com/libuv/libuv/pull/1346 and even more elegant, but I'm so indoctrinated to resist globals...

Comment (bzoz) [inline src/win/fs.c]: We keep function pointers as global values. This is similar, `uv__file_symlink_usermode_flag` is a global OS state, so I guess we are ok.

Also, there is no `flags` in `uv_loop_t`, so there is no other place to add this.

**Pull request closed without merging**